### ISSUE TITLE
Aligned alloc might not be defined

### ISSUE THREAD

**Issue by ETKNeil** [tensorflow/tensorflow#57706]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

2.0+

### Custom Code

No

### OS Platform and Distribution

Centos 6.0

### Python version

3.10

### Bazel version

5.2.0

### GCC/Compiler version

7.1.0

### Current Behaviour?

```shell
INFO: Analyzed target //tensorflow/lite/c:tensorflowlite_c (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /root/tensorflow-2.10.0/tensorflow/lite/BUILD:505:11: Compiling tensorflow/lite/interpreter_builder.cc failed: (Exit 1): gcc failed: error executing command /usr/local/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 47 arguments skipped)
tensorflow/lite/interpreter_builder.cc: In member function 'virtual void* tflite::{anonymous}::MallocDataAllocator::Allocate(size_t, size_t)':
tensorflow/lite/interpreter_builder.cc:312:12: error: 'aligned_alloc' was not declared in this scope
     return aligned_alloc(used_alignment, used_size);
            ^~~~~~~~~~~~~
Target //tensorflow/lite/c:tensorflowlite_c failed to build
```

Aligned alloc is not defined for a gcc compiler of version 7.1.0 (stdc11 and c++14):

```shell
[root]# gcc -dM -E - < /dev/null | grep STDC_VERSION
#define __STDC_VERSION__ 201112L

[root]# echo '#include<iostream>

int main() {
    if (__cplusplus == 201703L) std::cout << "C++17\n";
    else if (__cplusplus == 201402L) std::cout << "C++14\n";
    else if (__cplusplus == 201103L) std::cout << "C++11\n";
    else if (__cplusplus == 199711L) std::cout << "C++98\n";
    else std::cout << "pre-standard C++\n";
}' | g++ -x c++ -

[root]# ./a.out 
C++14

```

If we refer to [your code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/interpreter_builder.cc#L312) you need `TFLITE_USE_STD_ALIGNED_ALLOC` to be defined which happens with [all those conditions](https://github.com/tensorflow/tensorflow/blob/00a049661641ff13e0dd666fa692c97fb079e4bd/tensorflow/lite/interpreter_builder.cc#L49-L57)
```c
// aligned_alloc is available (via cstdlib/stdlib.h) with C++17/C11.
#if __cplusplus >= 201703L || __STDC_VERSION__ >= 201112L
#if !defined(__ANDROID__) || __ANDROID_API__ >= 28
// Neither Apple nor Windows provide aligned_alloc.
#if !defined(__APPLE__) && !defined(_WIN32)
#define TFLITE_USE_STD_ALIGNED_ALLOC
#endif
#endif
#endif
```

However having c11 is not enough, aligned_malloc was mainly shipped with c++17 so the first condition should have been `#if __cplusplus >= 201703L && __STDC_VERSION__ >= 201112L` and not a simple or.

[This](https://github.com/tensorflow/tensorflow/blob/4ea1236a02160e0c9bd8c3673cfae66dfb3f1b9b/tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc#L39) also need to change

### Standalone code to reproduce the issue

```shell
Just run the compilation with gcc 7.1.0
build -c opt //tensorflow/lite/c:tensorflowlite_c
```

### Relevant log output

```shell
INFO: Analyzed target //tensorflow/lite/c:tensorflowlite_c (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /root/tensorflow-2.10.0/tensorflow/lite/BUILD:505:11: Compiling tensorflow/lite/interpreter_builder.cc failed: (Exit 1): gcc failed: error executing command /usr/local/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 47 arguments skipped)
tensorflow/lite/interpreter_builder.cc: In member function 'virtual void* tflite::{anonymous}::MallocDataAllocator::Allocate(size_t, size_t)':
tensorflow/lite/interpreter_builder.cc:312:12: error: 'aligned_alloc' was not declared in this scope
     return aligned_alloc(used_alignment, used_size);
            ^~~~~~~~~~~~~
Target //tensorflow/lite/c:tensorflowlite_c failed to build
```
</details>

**Comment by ETKNeil** [comment]
I have been able to successfully compile with those two changes:
```shell
sed -i "s/201703L ||/201703L \&\&/" tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc
sed -i "s/201703L ||/201703L \&\&/" tensorflow/lite/interpreter_builder.cc
```

**Comment by mohantym** [comment]
Hi @ETKNeil !
Could you raise a PR from your end.
Thank you!

**Comment by pjpratik** [comment]
Hi @ETKNeil 

A new PR #61176 has been created as the original PR #57707 has been closed.

The issue will be closed once the PR is merged.

Thanks.
